src/cpu/flexcpu/flexcpu.cc

```diff
@@ -0,0 +1,1106 @@
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: Bradley Wang
+ */
+
+#include "cpu/flexcpu/flexcpu.hh"
+
+#include "arch/locked_mem.hh"
+#include "arch/mmapped_ipr.hh"
+#include "base/compiler.hh"
+#include "debug/FlexCPUCoreEvent.hh"
+
+using namespace std;
+using namespace TheISA;
+
+
+FlexCPU::FlexCPU(FlexCPUParams* params):
+    BaseCPU(params),
+    cacheBlockMask(~(cacheLineSize() - 1)),
+    executionLatency(params->execution_latency),
+    dataAddrTranslationUnit(this, params->clocked_dtb_translation, Cycles(0),
+                            0, name() + ".dtbUnit"),
+    executionUnit(this, params->clocked_execution, params->execution_latency,
+                  params->execution_bandwidth, name() + ".executionUnit"),
+    instFetchUnit(this, params->clocked_inst_fetch, 0, name() + ".iFetchUnit"),
+    instAddrTranslationUnit(this, params->clocked_itb_translation, Cycles(0),
+                            0, name() + ".itbUnit"),
+    issueUnit(this, params->clocked_issue, params->issue_latency,
+              params->issue_bandwidth, name() + ".issueUnit"),
+    memoryUnit(this, params->clocked_memory_request, Cycles(0),
+               name() + ".memoryUnit"),
+    _dataPort(name() + "._dataPort", this),
+    _instPort(name() + "._instPort", this)
+{
+    fatal_if(FullSystem, "FullSystem not implemented for FlexCPU");
+
+    for (ThreadID i = 0; i < numThreads; i++) {
+        // Move constructor + vector should allow us to avoid memory leaks like
+        // the SimpleCPU implementations have.
+        threads.push_back(m5::make_unique<FlexCPUThread>(this, i,
+            params->system, params->workload[i], params->itb, params->dtb,
+            params->isa[i],
+            params->fetch_buffer_size,
+            params->instruction_buffer_size,
+            params->strict_serialization));
+
+        threadContexts.push_back(threads[i]->getThreadContext());
+    }
+    // TODO add any other constructor details
+}
+
+FlexCPU::~FlexCPU()
```

Comment:
Not needed if empty. You *could* say `~Flexcpu() = default` if you want to explicitly use the default in the header file.

Reply:
done